### PR TITLE
update deploy

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: deployments/counter.js [hunk 1]
```diff
@@ -13,30 +13,38 @@ const { privateKey } = require('../privateKey');
...
-            unlockingTx.inputs[0].setScript(unlockingScript)
+            const unlockingTx = new bsv.Transaction();
+            
+            unlockingTx.addInput(createInputFromTx(prevTx))
+            .addOutput(new bsv.Transaction.Output({
+                script: newLockingScript,
+                satoshis: newAmount,
+              }))
+            .setInputScript(0, (self, output) => {
+                const preimage = getPreimage(self, output.script, output.satoshis)
```
Comment (xhliu) [inline deployments/counter.js:37]: change to
`getPreimage(tx, Output)`

File: helper.js [hunk 2]
```diff
@@ -287,6 +287,40 @@ const sleep = async(seconds) => {
...
+    satoshis: amountInContract,
+  }))
+  .change(address)
+  .sign(privateKey)
+  await sendTx(tx);
+  return tx;
+}
+
+function createInputFromTx(tx, outputIndex) {
+  return new bsv.Transaction.Input({
```
Comment (xhliu) [inline helper.js:317]: const outputIdx = outputIndex || 0

File: helper.js [hunk 3]
```diff
@@ -287,6 +287,40 @@ const sleep = async(seconds) => {
...
+    script: contract.lockingScript,
+    satoshis: amountInContract,
+  }))
+  .change(address)
+  .sign(privateKey)
+  await sendTx(tx);
+  return tx;
+}
+
+function createInputFromTx(tx, outputIndex) {
```
Comment (xhliu) [inline helper.js:316]: - Add comment: create an input spending from prevTx's output, with empty script

- createInputFromTx -> createInputFromPrevTx

File: helper.js [hunk 4]
```diff
@@ -287,6 +287,40 @@ const sleep = async(seconds) => {
...
+  return new bsv.Transaction.Input({
+    prevTxId: tx.id,
+    outputIndex: outputIndex || 0,
```
Comment (xhliu) [inline helper.js:319]: outputIndex: outputIdx,

File: helper.js [hunk 5]
```diff
@@ -287,6 +287,40 @@ const sleep = async(seconds) => {
...
+    script: new bsv.Script(), // placeholder
+    output: tx.outputs[ outputIndex || 0]
```
Comment (xhliu) [inline helper.js:321]: output: tx.outputs[outputIdx]

File: helper.js [hunk 6]
```diff
@@ -287,6 +287,40 @@ const sleep = async(seconds) => {
   })
 }
 
+async function deployContract(contract, amountInContract) {
```
Comment (xhliu) [inline helper.js:290]: amountInContract -> amount

File: deployments/demo.js [hunk 7]
```diff
@@ -12,17 +12,20 @@ const { privateKey } = require('../privateKey');
...
-        lockingTx.sign(privateKey)
-        const lockingTxid = await sendTx(lockingTx)
-        console.log('locking txid:     ', lockingTxid)
+        const tx =  await deployContract(demo, amount);
+        console.log('locking txid:     ', tx.id)
+
+        const unlockingTx = new bsv.Transaction();
+        unlockingTx.addInput(createInputFromTx(tx)) 
+        .change(privateKey.toAddress())
+        .setInputScript(0, (self, output) => {
```
Comment (xhliu) [inline deployments/demo.js:21]: (self, output)
->
(_, _)

**Review by xhliu — CHANGES_REQUESTED** [review]
👍  Much cleaner than before. A few changes to make it even cleaner.

File: deployments/counter.js (hunk 1 above)
Comment (zhfnjust) [reply deployments/counter.js]: fixed

File: helper.js (hunk 4 above)
Comment (zhfnjust) [reply helper.js]: fixed

File: helper.js (hunk 6 above)
Comment (zhfnjust) [reply helper.js]: fixed

File: helper.js (hunk 5 above)
Comment (zhfnjust) [reply helper.js]: fixed

File: helper.js (hunk 3 above)
Comment (zhfnjust) [reply helper.js]: fixed

File: testnet/counter.js [hunk 8]
```diff
@@ -13,30 +13,38 @@ const { privateKey } = require('../privateKey');
...
-            unlockingTx.inputs[0].setScript(unlockingScript)
+            const unlockingTx = new bsv.Transaction();
+            
+            unlockingTx.addInput(createInputFromTx(prevTx))
+            .addOutput(new bsv.Transaction.Output({
+                script: newLockingScript,
+                satoshis: newAmount,
+              }))
+            .setInputScript(0, (tx, output) => {
+                const preimage = getPreimage(tx, output.script, output.satoshis)
```
Comment (xhliu) [inline testnet/counter.js:37]: This requested change hasn't been made:

getPreimage(tx, Output)

File: testnet/counter.js [hunk 9]
```diff
@@ -13,30 +13,38 @@ const { privateKey } = require('../privateKey');
         const FEE = 1500;
```
Comment (xhliu) [inline testnet/counter.js:13]: Use dynamic feeperbyte everywhere, instead of hardcoding fee

File: testnet/advancedCounter.js [hunk 10]
```diff
@@ -59,34 +56,41 @@ function sleep(ms) {
...
+            0,
+            Signature.SIGHASH_ANYONECANPAY |
+            Signature.SIGHASH_SINGLE |
+            Signature.SIGHASH_FORKID
+          );
+          return advCounter
+            .increment(new SigHashPreimage(toHex(preimage)))
+            .toScript();
+        })
+        .sign(privateKey);
```
Comment (xhliu) [inline testnet/advancedCounter.js:84]: How does this sign every input, but input 0?

File: testnet/advancedTokenSale.js [hunk 11]
```diff
@@ -67,67 +67,72 @@ function sleep(ms) {
...
+          return advTokenSale.buy(
+            new SigHashPreimage(toHex(preimage)), // sighashPreimage
+            new Ripemd160(toHex(pkh)), // changePKH
+            tx.getChangeAmount(), // changeSats
+            new Bytes(toHex(publicKeys[i])), // buyer's public key
+            numBought // number of tokens purchased
+          ).toScript();
+
+        })
+        .seal()
```
Comment (xhliu) [inline testnet/advancedTokenSale.js:123]: What is this?

**Review by xhliu — CHANGES_REQUESTED** [review]
👍 Almost there.

Let us talk on how u handle the fee issue.

File: testnet/advancedCounter.js (hunk 10 above)
Comment (zhfnjust) [reply testnet/advancedCounter.js:84]: call sign for every input,  but return empty for contract input.

image

File: testnet/counter.js (hunk 8 above)
Comment (zhfnjust) [reply testnet/counter.js:37]: need to change getPreimage function

**Comment by xhliu** [comment]
add examples to set fee rate

File: testnet/advancedTokenSale.js (hunk 11 above)
Comment (zhfnjust) [reply testnet/advancedTokenSale.js:122]: seal Transaction to call callback automatically

File: testnet/advancedTokenSale.js [hunk 12]
```diff
@@ -118,7 +117,7 @@ function sleep(ms) {
             new Bytes(toHex(publicKeys[i])), // buyer's public key
             numBought // number of tokens purchased
           ).toScript();
-
+          
         })
         .seal()
         .sign(privateKey);
```
Comment (xhliu) [inline testnet/advancedTokenSale.js:123]: Shouldn't this be before seal()?

File: testnet/advancedCounter.js [hunk 13]
```diff
@@ -81,6 +81,9 @@ function sleep(ms) {
             .increment(new SigHashPreimage(toHex(preimage)))
             .toScript();
         })
+        .from(await fetchUtxos(privateKey.toAddress()))
+        .change(privateKey.toAddress())
+        .seal()
         .sign(privateKey);
```
Comment (xhliu) [inline testnet/advancedCounter.js:87]: Shouldn't this be before seal()?

File: testnet/advancedTokenSale.js (hunk 12 above)
Comment (zhfnjust) [reply testnet/advancedTokenSale.js:123]: sign should call after seal

File: deployments/demo.js (hunk 7 above)
Comment (zhfnjust) [reply deployments/demo.js]: can not be (_, _) , can be (_)

**Pull request merged**